src/types/redis_bloomfilter.cc

```diff
@@ -0,0 +1,378 @@
+  std::string bf_meta_bytes;
+  bf_metadata.Encode(&bf_meta_bytes);
+  batch->Put(metadata_cf_handle_, bf_meta_key, bf_meta_bytes);
+
+  return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status BloomFilter::bloomCheckAdd64(const Slice &bf_key, BFMetadata &bf_metadata, BloomHashval hashval,
+                                             ReadWriteMode mode, ObserverOrUniquePtr<rocksdb::WriteBatchBase> &batch,
+                                             int &ret) {
+  int found_unset = 0;
+  const uint64_t mod = (1LLU << bf_metadata.n2);
+  for (uint64_t i = 0; i < bf_metadata.hashes; i++) {
+    uint64_t x = ((hashval.a + i * hashval.b)) % mod;
+    uint32_t index = (x / kBloomFilterSegmentBits) * kBloomFilterSegmentBytes;
+    std::string sub_key, value;
+    InternalKey(bf_key, std::to_string(index), bf_metadata.version, storage_->IsSlotIdEncoded()).Encode(&sub_key);
+    rocksdb::Status s =
+        storage_->Get(rocksdb::ReadOptions(), sub_key, &value);  // todo: use bf_key in InternalKey maybe error
+    if (!s.ok() && !s.IsNotFound()) return s;
+    uint32_t byte_index = (x / 8) % kBloomFilterSegmentBytes;
+    if (byte_index >= value.size()) {  // expand the bloomfilter
+      size_t expand_size = 0;
+      if (byte_index >= value.size() * 2) {
+        expand_size = byte_index - value.size() + 1;
+      } else if (value.size() * 2 > kBloomFilterSegmentBytes) {
+        expand_size = kBloomFilterSegmentBytes - value.size();
+      } else {
+        expand_size = value.size();
+      }
+      value.append(expand_size, 0);
+    }
+
+    if (!testBitSetBit(value, x, mode)) {
+      if (mode == ReadWriteMode::MODE_READ) {
+        ret = 0;
+        return rocksdb::Status::OK();
+      } else {
+        found_unset = 1;
+        batch->Put(sub_key, value);
+        storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+        // todo: if the indexs are same, the value need to be updated immediately, if have other ways to compl it
+      }
+    }
+  }
+  if (mode == ReadWriteMode::MODE_READ) {
+    ret = 1;
+  } else {
+    ret = found_unset;
+  }
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status BloomFilter::Reserve(const Slice &user_key, double error_rate, uint64_t capacity, uint16_t expansion,
+                                     uint16_t scaling) {
+  std::string ns_key;
+  AppendNamespacePrefix(user_key, &ns_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  SBChainMetadata sb_chain_metadata;
+  rocksdb::Status s = getSBChainMetadata(ns_key, &sb_chain_metadata);
+  if (!s.ok() && !s.IsNotFound()) return s;
+  if (!s.IsNotFound()) {
+    return rocksdb::Status::InvalidArgument("item exists");
+  }
+
+  return createSBChain(ns_key, error_rate, capacity, expansion, scaling, sb_chain_metadata);
+}
+
+rocksdb::Status BloomFilter::Add(const Slice &user_key, const Slice &item, int &ret) {
```

Comment:
```suggestion
rocksdb::Status BloomFilter::Add(const Slice &user_key, const Slice &item, int* ret) {
```

Personally, I perfer `pointer` for output argument.